### PR TITLE
Canonicalize path before calling `startsWith`

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes Microsoft/vscode#52089 and #25279

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
What about test cases? Also, does this fix any of #25358 #24599 #25345 #25186?

**Comment by weswigham** [comment]
I do agree with Daniel that we probably should err on adding more test coverage tho, considering this is used for paths in declaration emit now, too.

File: src/compiler/moduleSpecifiers.ts [hunk 1]
```diff
@@ -290,7 +290,7 @@ namespace ts.moduleSpecifiers {
         const moduleSpecifier = getDirectoryOrExtensionlessFileName(moduleFileName);
         // Get a path that's relative to node_modules or the importing file's path
         // if node_modules folder is in this folder or any of its parent folders, no need to keep it.
-        if (!startsWith(sourceDirectory, moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex))) return undefined;
+        if (!startsWith(sourceDirectory, getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex)))) return undefined;
```
Comment (sheetalkamat) [inline src/compiler/moduleSpecifiers.ts]: Dont you need to do getCanonicalFileName for sourceDirectory as well?

**Comment by mhegazy** [comment]
Also looks related to https://github.com/Microsoft/TypeScript/issues/25338

File: src/compiler/moduleSpecifiers.ts (hunk 1 above)
Comment (andy-ms) [reply src/compiler/moduleSpecifiers.ts]: That comes from `importingSourceFile.path` which should already be canonicalized.

**Comment by andy-ms** [comment]
@sheetalkamat Do we have any existing case-insensitivity tests? I thought fourslash was case-insensitive but I can't reproduce the error in a fourslash test.

File: src/compiler/moduleSpecifiers.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/moduleSpecifiers.ts]: Makes sense. Good that you changed the type there so its easy to catch in future. Thanks.

**Comment by andy-ms** [comment]
Figured out why fourslash test wasn't working, I needed to put a capitalized name *before* the importing directory.

**Comment by mhegazy** [comment]
@andy-ms can you also verify that this change fixes #25358 #24599 #25345 and #25186?

**Comment by Kingwl** [comment]
image

seems fixed on my pc@20180705

**Comment by Kingwl** [comment]
> I'm running vscode 1.25.0 and typscript 2.9.2, still have this issue. Can anyone help? Thanks  
 1. install the typescript@next 
 2. set typescript tsdk to gloabl typescript lib, eg: `/usr/local/lib/node_modules/typescript/lib`

**Comment by smnbbrv** [comment]
This is all cool, but angular cli does not allow to use the newest typescript. What should we do?

**Comment by andy-ms** [comment]
@smnbbrv Just install typescript globally with `npm install -g typescript@next` and point the editor to that.

**Comment by smnbbrv** [comment]
@andy-ms thanks, it works!

**Comment by ratidzidziguri** [comment]
for me installing @next did not work

**Comment by ratidzidziguri** [comment]
solution for now was to only set VsCode to use workplace TS version. In my case that is the one installed with angular and it is 2.7.2

**Comment by smnbbrv** [comment]
Maybe I say bullshit but if the typescript version is so crucial why don’t you guys just bundle it with vscode or at least give users some hint about mismatching typescript versions and the consequences of that? The way how it is done right now is so counter intuitive; it was properly working one version before, then it gets broken and the only way to fix this is using beta versions of typescript; that’s kinda... weird

**Comment by mhegazy** [comment]
The next version of VSCode should include the fix. VSCode insiders tomorrow should include the fix. 

If you are on latest VSCode and want to try the fix out, you can find instructions on using a custom version of TS in your VSCode at: [Using Newer TypeScript Versions](https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions).

**Comment by no-stack-dub-sack** [comment]
@andy-ms  I just installed the latest insiders and am on TS 2.9.2 - you mentioned the new insiders should have the fix - should 2.9.2 + latest insiders be working? I'm still getting the same behavior as when I originally opened the issue.

**Comment by no-stack-dub-sack** [comment]
:point_up: never mind, it does work when I update to @next. So for those dealing with this issue that aren't using insiders and tsnext, will they have to wait until 3.0.0 is shipping with VSCode?

**Comment by lucasklaassen** [comment]
@ratidzidziguri can you explain what you mean by using a "work-space" version?

**Comment by mjgoethe** [comment]
@lucasklaassen or to anyone else looking: You can set your TS workspace version by following the steps [here](https://code.visualstudio.com/docs/languages/typescript#_using-newer-typescript-versions).

**Comment by alamothe** [comment]
This last release of VS Code is so, so buggy. 

Hope an upgrade is coming soon!

Does it make sense to allow people to revert after so many bugs were introduced?

**Comment by weswigham** [comment]
You can always set your workspace version of TS to an older (or newer!) version of TS.

**Comment by hobojoe** [comment]
> @ratidzidziguri can you explain what you mean by using a "work-space" version?


@lucasklaassen You can click on the right bottom toolbar of VS Code. Where you have the typescript version. From there you can choose the workspace version.

**Comment by djensen47** [comment]
Is there any chance this can get into a patch release? This apparently fixes #52089? Which is a pretty bad regression, it's loss of functionality. Basically auto imports import the wrong file. 

Thanks for all the hard work on this project. VS Code is awesome.

**Pull request merged**